lib/BuildSystem/BuildSystem.cpp

```diff
@@ -885,6 +887,123 @@ class DirectoryContentsTask : public Task {
 };
 
 
+/// This task is responsible for computing the filtered lists of files in
+/// directories.
+class FilteredDirectoryContentsTask : public Task {
+  std::string path;
+
+  /// The exclusion filters used while computing the signature
+  StringList filters;
+
+  /// The value for the input directory.
+  BuildValue directoryValue;
+
+  virtual void start(BuildEngine& engine) override {
+    // Depend on the node so that it can be connected up to potential producers.
+    engine.taskNeedsInput(
+        this, BuildKey::makeNode(path).toData(), /*inputID=*/0);
+    // Depend on the raw stat information, because we want to know if we need
+    // to rescan items under this if anything has changed.
```

Comment:
I'd comment here about the subtle dependency on linear scanning

Reply:
Actually, interestingly enough... in this particular case we could make it "non-subtle" by moving the additional input request into provideValue, although I don't think I'd actually recommend that at the moment.